src/Features/ExtendedTranslucency.cpp

```diff
@@ -0,0 +1,156 @@
+#include "ExtendedTranslucency.h"
+
+#include "../State.h"
+#include "../Util.h"
+
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(
+	ExtendedTranslucency::MaterialParams,
+	AlphaMode,
+	AlphaReduction,
+	AlphaSoftness);
+
+const RE::BSFixedString ExtendedTranslucency::NiExtraDataName = "AnisotropicAlphaMaterial";
+
+ExtendedTranslucency* ExtendedTranslucency::GetSingleton()
+{
+	static ExtendedTranslucency singleton;
+	return &singleton;
+}
+
+void ExtendedTranslucency::SetupResources()
+{
+	// Per material model settings for geometries with explicit material model
+	MaterialParams params{ 0, 0.f, 0.f, 0 };
+	for (int material = 0; material < MaterialModel::Max; material++) {
+		params.AlphaMode = material;
+		materialCB[material].emplace(params);
+	}
+	// Default material model buffer only changes in settings UI
+	materialDefaultCB.emplace(settings);
+}
+
+void ExtendedTranslucency::BSLightingShader_SetupGeometry(RE::BSRenderPass* pass)
+{
+	auto* transcluency = ExtendedTranslucency::GetSingleton();
+	static const REL::Relocation<const RE::NiRTTI*> NiIntegerExtraDataRTTI{ RE::NiIntegerExtraData::Ni_RTTI };
+
+	// TODO: OPTIMIZATION: Use materialCB[MaterialModel::Disabled] for geometry without NiAlphaProperty or Alpha Blend not enabled
+	ID3D11DeviceContext* context = State::GetSingleton()->context;
+	ID3D11Buffer* buffers[1];
+	if (auto* data = pass->geometry->GetExtraData(NiExtraDataName)) {
+		// Mods supporting this feature should adjust their alpha value in texture already
+		// And the texture should be adjusted based on full strength param
+		MaterialParams params = transcluency->settings;
+		if (data->GetRTTI() == NiIntegerExtraDataRTTI.get()) {
+			params.AlphaMode = std::clamp<int>(static_cast<RE::NiIntegerExtraData*>(data)->value, 0, MaterialModel::Max - 1);
+		} else {
+			params.AlphaMode = std::to_underlying(ExtendedTranslucency::MaterialModel::Default);
+		}
+
+		buffers[0] = materialCB[params.AlphaMode]->CB();
+		context->PSSetConstantBuffers(materialCBIndex, 1, buffers);
```

Comment:
I'm not a fan of spamming graphics commands, even if it doesn't cause a state switch.